pkg/client/messaging/client.go

```diff
@@ -141,108 +170,192 @@ func (c *Client) Services() []string {
-			return record, nil
+			conn = record
+			break
 		}
 	}
 
-	return nil, errConnForDIDNotFound
-}
-
-func (c *Client) sendToConnection(msg json.RawMessage, conn *connection.Record) error {
-	didCommMsg, err := service.ParseDIDCommMsgMap(msg)
-	if err != nil {
-		return err
+	if conn != nil {
+		return func() error {
+			return c.ctx.Messenger().Send(msg, conn.MyDID, conn.TheirDID)
+		}, nil
 	}
 
-	err = c.ctx.Messenger().Send(didCommMsg, conn.MyDID, conn.TheirDID)
+	dest, err := service.GetDestination(theirDID, c.ctx.VDRegistry())
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	return nil
+	return c.sendToDestination(msg, dest)
 }
 
-func (c *Client) sendToDestination(msg json.RawMessage, rqst *sendMsgOpts) error {
-	var dest *service.Destination
+func (c *Client) sendToDestination(msg service.DIDCommMsgMap, dest *service.Destination) (messageDispatcher, error) {
+	_, sigPubKey, err := c.ctx.KMS().CreateAndExportPubKeyBytes(kms.ED25519Type)
+	if err != nil {
+		return nil, err
+	}
 
-	// prepare destination
-	if rqst.theirDID != "" {
-		var err error
+	return func() error {
+		return c.ctx.Messenger().SendToDestination(msg, base58.Encode(sigPubKey), dest)
+	}, nil
+}
 
-		dest, err = service.GetDestination(rqst.theirDID, c.ctx.VDRegistry())
+func (c *Client) sendAndWaitForReply(ctx context.Context, action messageDispatcher, thID string,
+	replyType string) (json.RawMessage, error) {
+	var notificationCh chan NotificationPayload
+
+	if replyType != "" {
+		topic := uuid.New().String()
+		notificationCh = make(chan NotificationPayload)
+
+		err := c.msgRegistrar.Register(newMessageService(topic, replyType, nil,
+			NewNotifier(notificationCh, func(topic string, msgBytes []byte) bool {
+				var message struct {
+					Message service.DIDCommMsgMap `json:"message"`
+				}
+
+				err := json.Unmarshal(msgBytes, &message)
+				if err != nil {
+					logger.Debugf("failed to unmarshal incoming message reply: %s", err)
+					return false
+				}
+
+				msgThID, err := message.Message.ThreadID()
+				if err != nil {
+					logger.Debugf("failed to read incoming message reply thread ID: %s", err)
+					return false
+				}
+
+				return thID == "" || thID == msgThID
+			})))
 		if err != nil {
-			return err
+			return nil, err
 		}
-	} else if rqst.destination != nil {
-		dest = rqst.destination
-	}
 
-	if dest == nil {
-		return fmt.Errorf(errMsgDestinationMissing)
+		defer func() {
+			e := c.msgRegistrar.Unregister(topic)
+			if e != nil {
+				logger.Warnf("Failed to unregister wait for reply notifier: %w", e)
+			}
+		}()
 	}
 
-	_, sigPubKey, err := c.ctx.KMS().CreateAndExportPubKeyBytes(kms.ED25519Type)
+	err := action()
 	if err != nil {
-		return err
+		return nil, err
+	}
+
+	if notificationCh != nil {
+		return waitForResponse(ctx, notificationCh)
+	}
+
+	return json.RawMessage{}, nil
+}
+
+func waitForResponse(ctx context.Context, notificationCh chan NotificationPayload) (json.RawMessage, error) {
+	if ctx == nil {
```

Comment:
It's a bit weird to see this check (and default setting) so far down in the call stack.

Reply:
I can move it up

Reply:
Yes move it up and couple it with the await reply option. 

Also no "defaulting" at this layer - make them pass you a context in the API layer.

Reply:
Forced caller to pass context whenever await reply option is used.